### PR TITLE
#165343963 delete an article

### PR DESCRIPTION
**Pull request by Chiazokam**
#### Description
This PR creates the delete article endpoint

#### Type of change
- [x] Non-breaking change

#### How Has This Been Tested?
- [x] npm test
- [x] postman

image

#### Checklist:
- [x] write tests for the endpoint
- [x] create the delete endpoint
- [x] modify status code of author authentication

#### Story (PT)
[#165343963](https://www.pivotaltracker.com/story/show/165343963)

### CODE REVIEW

File: server/controllers/article.controller.js [hunk 1]
```diff
@@ -204,5 +204,25 @@ class ArticleController {
...
+   * @param {object} req
+   * @param {object} res
+   * @returns {object} response
+   */
+  static async deleteArticle(req, res) {
+    const {
+      params: { slug }, user: { id }
+    } = req;
+
+    const article = await Article.destroy({
```
Comment (davidokonji) [inline server/controllers/article.controller.js:325]: When deleting an article does the comments, threaded comments, and tags get deleted too? or just the article?

Comment (Chiazokam) [reply server/controllers/article.controller.js:325]: Will implement that right away

Comment (Chiazokam) [reply server/controllers/article.controller.js:325]: Done

Comment (BukkyOmo) [reply server/controllers/article.controller.js:325]: @davidokonji onDelete CASCADE should do that if it is being catered for in the model.

File: server/controllers/article.controller.js [hunk 2]
```diff
@@ -204,5 +204,25 @@ class ArticleController {
...
+      where: {
+        slug,
+        userId: id,
+      }
+    });
+    if (article) return Response(res, 200, `Successfully deleted ${article} article`);
+  }
```
Comment (BukkyOmo) [inline server/controllers/article.controller.js:333]: What happens when the article to be deleted is not in the database? Shouldn't there be an error message for that?

File: server/test/article.spec.js [hunk 3]
```diff
@@ -365,8 +368,48 @@ describe('Article email share', () => {
...
+
+  const validSlug = 'the-fattest-girl-3654677788';
+  const fakeSlug = 'David is so annoying it-12345678';
+
+  it('should check if the user is the author of the article', async () => {
+    const res = await chai.request(app)
+      .delete(`/api/v1/article/${validSlug}`)
+      .set({ Authorization: `Bearer ${userToken}` });
+    expect(res).to.have.status(401);
+    expect(res.body.message).to.eql('Action restricted to author of article');
```
Comment (chukwudidu) [inline server/test/article.spec.js:430]: Hi, @Chiazokam should an admin be given the privilege to delete an article?

Comment (Chiazokam) [reply server/test/article.spec.js:430]: No. Only the creator of an article can delete it. Admins should be able to suspend articles but that is another feature

Comment (chukwudidu) [reply server/test/article.spec.js:430]: @Chiazokam Alright

**Review by chukwudidu — APPROVED** [review]

File: server/routes/article.routes.js [hunk 4]
```diff
@@ -41,4 +46,6 @@ router.post('/comment/:commentId/like', [signInAuth, doesLikeExistInCommentForUs
 
 router.post('/:slug/share', [signInAuth, shareArticleCheck, checkArticleExist], tryCatch(shareArticle));
 
+router.delete('/:slug', [signInAuth, checkArticleExist, checkAuthor], tryCatch(deleteArticle));
```
Comment (davidokonji) [inline server/routes/article.routes.js]: From your routes, only the user can delete an article, then following RESTful conventions, the route for deleting a users article should be represented by `/users/article/: slug/`, what do you think?

File: server/controllers/article.controller.js [hunk 5]
```diff
@@ -204,5 +204,25 @@ class ArticleController {
...
+      params: { slug }, user: { id }
+    } = req;
+
+    const article = await Article.destroy({
+      where: {
+        slug,
+        userId: id,
+      }
+    });
+    if (article) return Response(res, 200, `Successfully deleted ${article} article`);
```
Comment (geekdanitek) [inline server/controllers/article.controller.js:331]: There should a check if the article was not deleted successfully.

**Review by geekdanitek — CHANGES_REQUESTED** [review]

File: server/test/article.spec.js [hunk 6]
```diff
@@ -365,8 +368,48 @@ describe('Article email share', () => {
...
+    const userResponse = await chai
+      .request(app)
+      .post('/api/v1/auth/login')
+      .send({
+        email: 's.david@example.com',
+        password: 'soldier'
+      });
+
+    authorToken = userResponse.body.data.token;
+  });
```
Comment (davidokonji) [inline server/test/article.spec.js]: Looking through the codebase, there is already a before call made, why did you make multiple calls to the same login route? refactoring the first one to suit your need could be better

File: server/controllers/article.controller.js (hunk 5 above)
Comment (Chiazokam) [reply server/controllers/article.controller.js:331]: Alright...will put that

Comment (Chiazokam) [reply server/controllers/article.controller.js:331]: Done

**Review by geekdanitek — APPROVED** [review]

File: server/routes/article.routes.js (hunk 4 above)
Comment (Chiazokam) [reply server/routes/article.routes.js]: I don't think so. Every other user operation on his article in the codebase does not follow the `users/article` route. The `users` in the feature definition only refers to the fact that only a user can delete his article.

Comment (davidokonji) [reply server/routes/article.routes.js]: @Chiazokam if the user is to delete an article it should be related to a user and not an article,  performing a delete user article with `/articles/: slug` won't be the best fit for deleting an article that is peculiar to a user

File: server/test/article.spec.js (hunk 6 above)
Comment (Chiazokam) [reply server/test/article.spec.js]: Will work on this

Comment (Chiazokam) [reply server/test/article.spec.js]: Done

File: server/routes/article.routes.js (hunk 4 above)
Comment (Chiazokam) [reply server/routes/article.routes.js]: Great idea but the rubrics specified so `DELETE /api/articles/:slug`

File: server/controllers/article.controller.js [hunk 7]
```diff
@@ -204,5 +204,26 @@ class ArticleController {
...
+   * @param {object} req
+   * @param {object} res
+   * @returns {object} response
+   */
+  static async deleteArticle(req, res) {
+    const {
+      params: { slug }, user: { id }
+    } = req;
+
+    const article = await Article.destroy({
```
Comment (davidokonji) [inline server/controllers/article.controller.js:325]: @Chiazokam for articles that include images, does this endpoint delete article images on cloudinary? if it does not, It is necessary to delete images alongside the articles. if you need help with that reach out to me.

**Review by davidokonji — CHANGES_REQUESTED** [review]

File: server/controllers/article.controller.js (hunk 7 above)
Comment (Chiazokam) [reply server/controllers/article.controller.js:325]: Did not take care of that. Thanks for drawing my attention

Comment (Chiazokam) [reply server/controllers/article.controller.js:325]: Done

**Review by davidokonji — APPROVED** [review]
Good work done here

File: server/utils/articles/deleteImage.util.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    where: {
+      slug,
+    }
+  });
+  const imageUrl = article.dataValues.image;
+  if (!imageUrl) return next();
+  const secureImageId = imageUrl.slice(70, 90);
+  cloudinary.v2.uploader.destroy(secureImageId);
+  return next();
+};
```
Comment (TheDrizzyWay) [inline server/utils/articles/deleteImage.util.js:19]: A little clarification please. Why 70 and 90?

Comment (Chiazokam) [reply server/utils/articles/deleteImage.util.js:19]: I had to splice the image url in the db to get the image id which is on cloudinary. 70 and 90 are the start and end of the string I need.
This is what is deleted from cloudinary

Comment (TheDrizzyWay) [reply server/utils/articles/deleteImage.util.js:19]: Got it.

File: server/controllers/article.controller.js [hunk 9]
```diff
@@ -204,5 +204,26 @@ class ArticleController {
...
+      where: {
+        slug,
+        userId: id,
+      }
+    });
+    if (article) return Response(res, 200, `Successfully deleted ${article} article`);
```
Comment (samorahno) [inline server/controllers/article.controller.js:331]: Good job @Chiazokam, I think this line will be okay without returning the count of articles deleted that is `Successfully deleted article`, or I will need clarification if this endpoint can handle multiple delete. what do you think about this?

**Review by samorahno — CHANGES_REQUESTED** [review]

**Review by TheDrizzyWay — APPROVED** [review]

File: server/controllers/article.controller.js (hunk 9 above)
Comment (Chiazokam) [reply server/controllers/article.controller.js:331]: I thought it wise to give the user an idea of what happened behind, hence the count

File: server/controllers/article.controller.js (hunk 2 above)
Comment (Chiazokam) [reply server/controllers/article.controller.js:333]: Yes, there's a check as a middleware to check if the article to be deleted exists

**Review by samorahno — APPROVED** [review]

**Review by paradisekelechi — APPROVED** [review]

**Pull request merged**